File: sdks/typescript/src/types/documents.ts
/**
 * Document types.
 *
 * @module types/documents
 * @see edgequake/crates/edgequake-api/src/handlers/documents_types.rs
 */

import type { Timestamp } from "./common.js";

// ── Upload ────────────────────────────────────────────────────

export interface UploadDocumentRequest {
  content: string;
  title?: string;
  metadata?: Record<string, unknown>;
  async_processing?: boolean;
  track_id?: string;
  enable_gleaning?: boolean;
  max_gleaning?: number;
  use_llm_summarization?: boolean;
}

export interface UploadDocumentResponse {
  document_id: string;
  status: string;
  track_id?: string;
  message?: string;
}

// ── File Upload ───────────────────────────────────────────────

export interface UploadFileMetadata {
  title?: string;
  metadata?: Record<string, unknown>;
}

export interface BatchUploadResponse {
  documents: UploadDocumentResponse[];
  total: number;
  succeeded: number;
  failed: number;
}

/** Response from single file upload. */
export type UploadFileResponse = UploadDocumentResponse;

// ── List ──────────────────────────────────────────────────────

export interface ListDocumentsQuery {
  limit?: number;
  offset?: number;
  status?: string;
  search?: string;
}

export interface DocumentInfo {
  document_id: string;
  title?: string;
  status: string;
  created_at: Timestamp;
  updated_at?: Timestamp;
  chunk_count?: number;
  entity_count?: number;
  content_length?: number;
}

export interface DocumentDetail extends DocumentInfo {
  content?: string;
  metadata?: Record<string, unknown>;
  track_id?: string;
  error_message?: string;
}

// ── Track ─────────────────────────────────────────────────────

export interface TrackStatusResponse {
  track_id: string;
  status: string;
  progress?: number;
  documents?: Array<{
    document_id: string;
    status: string;
    error?: string;
  }>;
  created_at?: Timestamp;
  updated_at?: Timestamp;
}

// ── Scan ──────────────────────────────────────────────────────

export interface ScanDirectoryRequest {
  path: string;
  recursive?: boolean;
  max_files?: number;
  extensions?: string[];
}

export interface ScanDirectoryResponse {
  total_files: number;
  queued: number;
  skipped: number;
  track_id: string;
}

// ── Reprocess ─────────────────────────────────────────────────

export interface ReprocessRequest {
  max_reprocess?: number;
}

export interface ReprocessResponse {
  reprocessed: number;
  message: string;
}

export interface RecoverStuckRequest {
  stuck_threshold_minutes?: number;
}

export interface RecoverStuckResponse {
  recovered: number;
  message: string;
}

// ── Deletion Impact ───────────────────────────────────────────

export interface DeletionImpactResponse {
  document_id: string;
  chunks_affected: number;
  entities_affected: number;
  relationships_affected: number;
}

// ── Chunks ────────────────────────────────────────────────────

export interface RetryChunksResponse {
  retried: number;
  message: string;
}

export interface FailedChunkInfo {
  chunk_id: string;
  error: string;
  created_at: Timestamp;
}

export interface FailedChunksResponse {
  chunks: FailedChunkInfo[];
  total: number;
}

// ── PDF ───────────────────────────────────────────────────────

export interface PdfUploadMetadata {
  title?: string;
  metadata?: Record<string, unknown>;
}

export interface PdfUploadResponse {
  pdf_id: string;
  document_id?: string;
  status: string;
  track_id: string;
  message?: string;
}

export interface ListPdfsQuery {
  limit?: number;
  offset?: number;
  status?: string;
}

export interface PdfInfo {
  pdf_id: string;
  document_id?: string;
  filename: string;
  status: string;
  file_size: number;
  page_count?: number;
  created_at: Timestamp;
}

export interface PdfStatusResponse extends PdfInfo {
  error_message?: string;
  markdown_content?: string;
  track_id?: string;
}

export interface PdfContentResponse {
  pdf_id: string;
  markdown: string;
  page_count: number;
}

export interface PdfProgressResponse {
  track_id: string;
  status: string;
  progress: number;
  current_page?: number;
  total_pages?: number;
  message?: string;
}

export interface PdfRetryResponse {
  pdf_id: string;
  status: string;
  message: string;
}

// ── Delete All ────────────────────────────────────────────────

export interface DeleteAllResponse {
  deleted: number;
  message: string;
}

File: sdks/rust/src/types/documents.rs
//! Document-related types.

use serde::{Deserialize, Serialize};

use super::common::PaginationInfo;

/// Response from uploading a document.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct UploadDocumentResponse {
    pub id: String,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    pub track_id: Option<String>,
}

/// Document summary in list responses.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct DocumentSummary {
    pub id: String,
    #[serde(default)]
    pub file_name: Option<String>,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub file_size: Option<u64>,
    #[serde(default)]
    pub mime_type: Option<String>,
    #[serde(default)]
    pub entity_count: Option<u32>,
    #[serde(default)]
    pub chunk_count: Option<u32>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
}

/// Response from listing documents.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ListDocumentsResponse {
    #[serde(default)]
    pub documents: Vec<DocumentSummary>,
    #[serde(default)]
    pub pagination: Option<PaginationInfo>,
}

/// Response from tracking document processing status.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct TrackStatusResponse {
    pub track_id: String,
    pub status: String,
    #[serde(default)]
    pub progress: Option<f64>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    pub document_id: Option<String>,
}

/// Response from directory scanning.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScanResponse {
    #[serde(default)]
    pub files_found: u32,
    #[serde(default)]
    pub files_queued: u32,
    #[serde(default)]
    pub files_skipped: u32,
}

/// Response from deletion impact analysis.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct DeletionImpactResponse {
    #[serde(default)]
    pub entity_count: u32,
    #[serde(default)]
    pub relationship_count: u32,
    #[serde(default)]
    pub chunk_count: u32,
}

/// PDF upload response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfUploadResponse {
    pub id: String,
    #[serde(default)]
    pub status: Option<String>,
    #[serde(default)]
    pub track_id: Option<String>,
}

/// PDF progress response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfProgressResponse {
    pub track_id: String,
    pub status: String,
    #[serde(default)]
    pub progress: Option<f64>,
}

/// PDF content response.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PdfContentResponse {
    pub id: String,
    #[serde(default)]
    pub markdown: Option<String>,
}

/// Scan request parameters.
#[derive(Debug, Clone, Serialize)]
pub struct ScanRequest {
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub recursive: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub extensions: Option<Vec<String>>,
}
